tests/compas/files/test_xml.py

```diff
@@ -59,3 +64,73 @@ def test_default_namespace_to_string():
     assert b'xmlns="https://default.org/namespace"' in xml_string
     assert b'<xacro:bamboo' in xml_string or b'<ns1:bamboo' in xml_string
     assert b'<robot' in xml_string or b'<ns0:robot' in xml_string
+
+
+def test_nested_default_namespaces():
+    xml = XML.from_string(
+        """<?xml version="1.0"?>
+        <main xmlns="https://ita.arch.ethz.ch/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="test-xml">
+            <item name="item1"><subitem /></item>
+            <item xmlns="https://ethz.ch" name="item2"><subitem /></item>
+        </main>"""
+    )
+
+    assert xml.root.attrib['xmlns'] == 'https://ita.arch.ethz.ch/'
+    assert xml.root.attrib['xmlns:xsi'] == 'http://www.w3.org/2001/XMLSchema-instance'
+
+    # first element redefines default namespace
+    assert list(xml.root)[1].attrib['xmlns'] == 'https://ethz.ch'
+    assert list(xml.root)[1].attrib['name'] == 'item2'
+
+
+# This is the same test as above, but the code paths of loading from file vs from string are different on pre-3.8 cpython
+def test_nested_default_namespaces_from_file(sample_with_nested_default_ns):
+    xml = XML.from_file(sample_with_nested_default_ns)
+
+    assert xml.root.attrib['xmlns'] == 'https://ita.arch.ethz.ch/'
+    assert xml.root.attrib['xmlns:xsi'] == 'http://www.w3.org/2001/XMLSchema-instance'
+
+    # first element redefines default namespace
+    assert list(xml.root)[1].attrib['xmlns'] == 'https://ethz.ch'
+    assert list(xml.root)[1].attrib['name'] == 'item2'
+
+
+def test_no_root_default_namespace():
+    xml = XML.from_string(
+        """<?xml version="1.0"?>
+        <main name="test-xml">
+            <item name="item1"><subitem /></item>
+            <item xmlns="https://ethz.ch" name="item2"><subitem /></item>
+        </main>"""
+    )
+
+    assert not xml.root.attrib.get('xmlns')
+    assert list(xml.root)[1].attrib['xmlns'] == 'https://ethz.ch'
+    assert list(xml.root)[1].attrib['name'] == 'item2'
+
+
+def test_no_default_namespace():
+    xml = XML.from_string(
+        """<?xml version="1.0"?><main name="test-xml"></main>"""
+    )
+
+    assert not xml.root.attrib.get('xmlns')
+    assert xml.root.attrib['name'] == 'test-xml'
+
+
+def test_namespace_expansion():
+    xml = XML.from_string(
```

Comment:
would it be worthwhile to have a similar test but `from_file`?

Reply:
added in https://github.com/compas-dev/compas/pull/686/commits/35b96c5a956a3c02882f6242c913a43f78875b68